captum/attr/_core/deep_lift.py

```diff
@@ -656,35 +680,48 @@ def _compute_mean_across_baselines(self, inp_bsz, base_bsz, attribution):
         return torch.mean(attribution.view(attr_shape), axis=1, keepdim=False)
 
 
-def nonlinear(module, delta_in, delta_out, grad_input, grad_output, eps=1e-10):
+def nonlinear(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
     r"""
     grad_input: (dLoss / dprev_layer_out, dLoss / wij, dLoss / bij)
     grad_output: (dLoss / dlayer_out)
     https://github.com/pytorch/pytorch/issues/12331
     """
+    delta_in, delta_out = _compute_diffs(inputs, outputs)
+
+    new_grad_inp = list(grad_input)
+
     # supported non-linear modules take only single tensor as input hence accessing
     # only the first element in `grad_input` and `grad_output`
-    grad_input[0] = torch.where(
-        abs(delta_in[0]) < eps,
-        grad_input[0],
-        grad_output[0] * delta_out[0] / delta_in[0],
+    new_grad_inp[0] = torch.where(
+        abs(delta_in) < eps, new_grad_inp[0], grad_output[0] * delta_out / delta_in,
     )
-    return grad_input
+    return new_grad_inp
+
 
+def softmax(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
+    delta_in, delta_out = _compute_diffs(inputs, outputs)
 
-def softmax(module, delta_in, delta_out, grad_input, grad_output, eps=1e-10):
+    new_grad_inp = list(grad_input)
     grad_input_unnorm = torch.where(
-        abs(delta_in[0]) < eps,
-        grad_input[0],
-        grad_output[0] * delta_out[0] / delta_in[0],
+        abs(delta_in) < eps, new_grad_inp[0], grad_output[0] * delta_out / delta_in,
     )
     # normalizing
     n = np.prod(grad_input[0].shape)
-    grad_input[0] = grad_input_unnorm - grad_input_unnorm.sum() * 1 / n
-    return grad_input
 
+    # updating only the first half
+    new_grad_inp[0] = grad_input_unnorm - grad_input_unnorm.sum() * 1 / n
+    return new_grad_inp
+
+
+def maxpool1d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
+    input, input_ref = inputs.chunk(2)
+    output, output_ref = outputs.chunk(2)
+
+    delta_in = input - input_ref
+    delta_in = torch.cat(2 * [delta_in])
+    delta_out_xmax = torch.max(output, output_ref)
+    delta_out = torch.cat([delta_out_xmax - output_ref, output - delta_out_xmax])
```

Comment:
I'll add more description but this logic is based on the logic from original SHAP implementation.

Reply:
Yeah, it is basically taking cross maximum between delta outputs for the inputs and corresponding references. It relies on the gradients for the reference values rather than the actual inputs because they are able to describe the differences and contributions better. 

I added test cases with large reference values and made sure that in the majority of the cases it is relying on the references and that it converges properly based on the completeness property.